File: mobile/ios/packages/town-crier-domain/Sources/Entities/SubscriptionTier.swift
/// Subscription level determining feature access.
///
/// Cases are ordered from lowest to highest entitlement so that
/// `Comparable` picks the richer tier when merging multiple sources
/// (e.g. server profile vs StoreKit).
public enum SubscriptionTier: String, Equatable, Hashable, Sendable, Comparable {
  case free
  case personal
  case pro

  // MARK: - Comparable

  private var rank: Int {
    switch self {
    case .free:
      0
    case .personal:
      1
    case .pro:
      2
    }
  }

  public static func < (lhs: SubscriptionTier, rhs: SubscriptionTier) -> Bool {
    lhs.rank < rhs.rank
  }
}

File: mobile/android/domain/src/main/kotlin/uk/towncrierapp/domain/subscriptions/SubscriptionTier.kt
package uk.towncrierapp.domain.subscriptions

/**
 * Subscription level determining feature access. Declaration order is
 * significant: enum ordinal gives `Comparable` for free, so `maxOf(a, b)`
 * always picks the richer tier when merging multiple sources (server
 * profile, cached value, JWT claim — see [resolveTier]).
 */
public enum class SubscriptionTier(
    public val wireValue: String,
) {
    FREE("Free"),
    PERSONAL("Personal"),
    PRO("Pro"),
    ;

    public companion object {
        /**
         * Parses a tier from its wire string. Case-insensitive so it accepts
         * both the server's PascalCase vocabulary ("Personal") and the JWT
         * `subscription_tier` claim's lowercase spelling ("personal").
         * Returns `null` for anything unrecognised — callers decide the
         * fallback (never silently default to FREE here, see tc-exg6).
         */
        public fun fromWireValue(value: String): SubscriptionTier? = entries.firstOrNull { it.wireValue.equals(value, ignoreCase = true) }
    }
}
